### ISSUE TITLE
The experimental results of the current version of the code are inconsistent with the paper.

### ISSUE THREAD

**Issue by default-user-tmp** [songleiW/SecGNN#2]
I cloned the code from github and ran it without modification, but the experimental results don't look good. 

epoch:201		error: -5.86991e+07
Accuracy of validation set: 0.154
Online comm: 22315 MB
Offline comm: 32600 MB
Training time: 127.192 minutes
Offline Time: 5.43333 minutes

Could you please update to the latest experimental code?

**Comment by Nith077** [comment]
> I cloned the code from github and ran it without modification, but the experimental results don't look good.
> 
> epoch:201 error: -5.86991e+07 Accuracy of validation set: 0.154 Online comm: 22315 MB Offline comm: 32600 MB Training time: 127.192 minutes Offline Time: 5.43333 minutes
> 
> Could you please update to the latest experimental code?


Could you please help me in giving the steps for compiling the whole github code because I am struggling in building it....

**Comment by songleiW** [comment]
I apologize for any trouble you encountered. I've updated the code with the unclear version. If that doesn't work, you can use our original uncleaned code at this link: https://github.com/songleiW/encryptGCN-unclear, which runs with one-click and requires no modifications. If the experimental results are inconsistent, please try running it on macOS. Below are the results from a one-click local run on my Mac; please note that different systems (32-bit or 64-bit) may also impact the results. Feel free to ask any additional questions!

Training: 
[Training.pdf](https://github.com/user-attachments/files/17412024/Training.pdf)


Inference: 
[Inference.pdf](https://github.com/user-attachments/files/17412025/Inference.pdf)

**Comment by SreeHariNathanKS-BTechCCE** [comment]
> I apologize for any trouble you encountered. I've updated the code with the unclear version. If that doesn't work, you can use our original uncleaned code at this link: https://github.com/songleiW/encryptGCN-unclear, which runs with one-click and requires no modifications. If the experimental results are inconsistent, please try running it on macOS. Below are the results from a one-click local run on my Mac; please note that different systems (32-bit or 64-bit) may also impact the results. Feel free to ask any additional questions!
> 
> Training: [Training.pdf](https://github.com/user-attachments/files/17412024/Training.pdf)
> 
> Inference: [Inference.pdf](https://github.com/user-attachments/files/17412025/Inference.pdf)



Sir, we implemented this unclear version code in MacOS.

System Details: 
      PC Model: MacBookPro
      Processor: M1 Pro (12-core CPU, 16-core GPU, 32-core Neural Engine)
      RAM: 16GB

But, still we are getting an accuracy of 0.16. I am not sure on how to proceed with this code. Please list out the dependencies required as well. 

So, can you please help us to sort out this issue with the code sir? Thank you

**Comment by Z-Patar** [comment]
> Sir, we implemented this unclear version code in MacOS.
> 
> System Details: PC Model: MacBookPro Processor: M1 Pro (12-core CPU, 16-core GPU, 32-core Neural Engine) RAM: 16GB
> 
> But, still we are getting an accuracy of 0.16. I am not sure on how to proceed with this code. Please list out the dependencies required as well.
> 
> So, can you please help us to sort out this issue with the code sir? Thank you

Hello, I encountered the same problem when trying to reproduce this experiment. 

Also, when running the code from the SecGNN repository, I noticed that its error value oscillates back and forth between magnitudes of positive and negative e+07 or higher. 

Could you please tell me how you solved this problem? This is very important to me. 

I would be very grateful if you could help me.

**Comment by songleiW** [comment]
> > Sir, we implemented this unclear version code in MacOS.
> > System Details: PC Model: MacBookPro Processor: M1 Pro (12-core CPU, 16-core GPU, 32-core Neural Engine) RAM: 16GB
> > But, still we are getting an accuracy of 0.16. I am not sure on how to proceed with this code. Please list out the dependencies required as well.
> > So, can you please help us to sort out this issue with the code sir? Thank you
> 
> Hello, I encountered the same problem when trying to reproduce this experiment.
> 
> Also, when running the code from the SecGNN repository, I noticed that its error value oscillates back and forth between magnitudes of positive and negative e+07 or higher.
> 
> Could you please tell me how you solved this problem? This is very important to me.
> 
> I would be very grateful if you could help me.

You can access our original, uncleaned implementation at https://github.com/songleiW/encryptGCN-unclear. It can be compiled and executed directly using CLion, requiring no code modifications. All dependencies are standard libraries, so no special installations or configurations are needed.

**Comment by Z-Patar** [comment]
> You can access our original, uncleaned implementation at https://github.com/songleiW/encryptGCN-unclear. It can be compiled and executed directly using CLion, requiring no code modifications. All dependencies are standard libraries, so no special installations or configurations are needed.

你好，我在Ubuntu24.04中使用cmake编译并执行了 https://github.com/songleiW/encryptGCN-unclear 仓库中的代码，但准确率依旧只有0.158。运行结果如下，能请教为什么吗

> user@user:/wslProject/encryptGCN-unclear-main/build$ ./encryptGCN
> ---随机数传输完成---
> ---密文数据读取完成---
> ---密文模型读取完成---
> ---初始化完成---
> ---准确率：0.158---
> Online comm: 368 MB
> Offline comm: 551 MB
> Inference time: 0.357954 minutes
> Offline Time: 0.0918333 minutes

**Comment by songleiW** [comment]
> > You can access our original, uncleaned implementation at https://github.com/songleiW/encryptGCN-unclear. It can be compiled and executed directly using CLion, requiring no code modifications. All dependencies are standard libraries, so no special installations or configurations are needed.
> 
> 你好，我在Ubuntu24.04中使用cmake编译并执行了 https://github.com/songleiW/encryptGCN-unclear 仓库中的代码，但准确率依旧只有0.158。运行结果如下，能请教为什么吗
> 
> > user@user:/wslProject/encryptGCN-unclear-main/build$ ./encryptGCN
> > ---随机数传输完成---
> > ---密文数据读取完成---
> > ---密文模型读取完成---
> > ---初始化完成---
> > ---准确率：0.158---
> > Online comm: 368 MB
> > Offline comm: 551 MB
> > Inference time: 0.357954 minutes
> > Offline Time: 0.0918333 minutes

同学，你试一下重新训练，然后用新训练的模型再做inference，因为不同机器训练的精度不一样，直接用我这边训练好的模型不一定和你的机器匹配。还有要注意你的机器的位数，是32位的还是64位的。附件是我换了一台电脑MacBook pro重新跑的结果。如果在Ubuntu上跑不出来，可以尝试一下用Mac跑。

希望能够帮到你，期待你的反馈。

[训练结果.pdf](https://github.com/user-attachments/files/23247881/default.pdf)

**Comment by Z-Patar** [comment]
> 同学，你试一下重新训练，然后用新训练的模型再做inference，因为不同机器训练的精度不一样，直接用我这边训练好的模型不一定和你的机器匹配。还有要注意你的机器的位数，是32位的还是64位的。附件是我换了一台电脑MacBook pro重新跑的结果。如果在Ubuntu上跑不出来，可以尝试一下用Mac跑。
> 
> 希望能够帮到你，期待你的反馈。
> 
> [训练结果.pdf](https://github.com/user-attachments/files/23247881/default.pdf)

非常感谢你的帮助，我的机器是win11下的wsl，版本为Ubuntu24.04，64位。并且受条件所限，无法使用Mac测试代码。所以无法确认是否是因为Mac机器与wsl存在差异导致我们的结果出现差异。

照你所说，我尝试重新训练再inference（将`main.cpp > main()`中的`if(0)`条件改为`1`编译后训练，再将其改回`0`并编译后进行inference）。但准确率依旧只有0.194。

我的训练过程和inference过程如下。

Training：[Training.pdf](https://github.com/user-attachments/files/23280154/Training.pdf)

Inference：[Inference.pdf](https://github.com/user-attachments/files/23280155/Inference.pdf)

在训练过程中我观察到error在一个很大的数量级反复震荡，没有按照预期收敛并下降，但这个现象的原因我暂时没有发现。
